File: nobodywho/kotlin/common/src/TextToSpeech.kt
package ai.nobodywho

import java.io.Closeable
import uniffi.nobodywho.RustTextToSpeech as InternalTts
import uniffi.nobodywho.loadTextToSpeech as loadInternalTextToSpeech

enum class TextToSpeechArchitecture(internal val value: String) {
    KOKORO("kokoro"),
    POCKET_TTS("pocket-tts"),
    SUPERTONIC("supertonic")
}

enum class TextToSpeechDevice(internal val value: String) {
    AUTO("auto"),
    CPU("cpu"),
    CUDA("cuda")
}

/** Text-to-speech synthesizer that returns WAV bytes. */
class TextToSpeech private constructor(
    private val inner: InternalTts
) : Closeable {
    /** Create a TextToSpeech synthesizer synchronously. */
    constructor(
        source: String,
        architecture: TextToSpeechArchitecture? = null,
        voice: String? = null,
        language: String? = null,
        speed: Float? = null,
        steps: UInt? = null,
        silenceDuration: Float? = null,
        precision: String? = null,
        temperature: Float? = null,
        huggingfaceToken: String? = null,
        device: TextToSpeechDevice = TextToSpeechDevice.AUTO
    ) : this(
        InternalTts(
            source = source,
            architecture = architecture?.value,
            voice = voice,
            language = language,
            speed = speed,
            steps = steps,
            silenceDuration = silenceDuration,
            precision = precision,
            temperature = temperature,
            huggingfaceToken = huggingfaceToken,
            device = device.value
        )
    )

    companion object {
        /** Create a TextToSpeech synthesizer asynchronously. */
        suspend fun load(
            source: String,
            architecture: TextToSpeechArchitecture? = null,
            voice: String? = null,
            language: String? = null,
            speed: Float? = null,
            steps: UInt? = null,
            silenceDuration: Float? = null,
            precision: String? = null,
            temperature: Float? = null,
            huggingfaceToken: String? = null,
            device: TextToSpeechDevice = TextToSpeechDevice.AUTO
        ): TextToSpeech = TextToSpeech(
            loadInternalTextToSpeech(
                source = source,
                architecture = architecture?.value,
                voice = voice,
                language = language,
                speed = speed,
                steps = steps,
                silenceDuration = silenceDuration,
                precision = precision,
                temperature = temperature,
                huggingfaceToken = huggingfaceToken,
                device = device.value
            )
        )
    }

    /** Synthesize text and return WAV bytes. */
    suspend fun synthesize(text: String): ByteArray = inner.synthesizeAsync(text)

    /** Synthesize text synchronously and return WAV bytes. */
    fun synthesizeSync(text: String): ByteArray = inner.synthesize(text)

    /** Free the underlying Rust resources. */
    fun destroy() = inner.destroy()
    override fun close() { destroy() }
}

File: nobodywho/swift/src/TextToSpeech.swift
import Foundation
import NobodyWhoGenerated

public enum TextToSpeechArchitecture: String {
    case kokoro
    case pocketTts = "pocket-tts"
    case supertonic
}

public enum TextToSpeechDevice: String {
    case auto
    case cpu
    case cuda
}

/// Text-to-speech synthesizer that returns WAV bytes.
public class TextToSpeech {
    private let inner: NobodyWhoGenerated.RustTextToSpeech

    /// Create a TextToSpeech synthesizer synchronously.
    public init(
        source: String,
        architecture: TextToSpeechArchitecture? = nil,
        voice: String? = nil,
        language: String? = nil,
        speed: Float? = nil,
        steps: UInt32? = nil,
        silenceDuration: Float? = nil,
        precision: String? = nil,
        temperature: Float? = nil,
        huggingfaceToken: String? = nil,
        device: TextToSpeechDevice = .auto
    ) throws {
        self.inner = try NobodyWhoGenerated.RustTextToSpeech(
            source: source,
            architecture: architecture?.rawValue,
            voice: voice,
            language: language,
            speed: speed,
            steps: steps,
            silenceDuration: silenceDuration,
            precision: precision,
            temperature: temperature,
            huggingfaceToken: huggingfaceToken,
            device: device.rawValue
        )
    }

    private init(inner: NobodyWhoGenerated.RustTextToSpeech) {
        self.inner = inner
    }

    /// Create a TextToSpeech synthesizer asynchronously.
    public static func load(
        source: String,
        architecture: TextToSpeechArchitecture? = nil,
        voice: String? = nil,
        language: String? = nil,
        speed: Float? = nil,
        steps: UInt32? = nil,
        silenceDuration: Float? = nil,
        precision: String? = nil,
        temperature: Float? = nil,
        huggingfaceToken: String? = nil,
        device: TextToSpeechDevice = .auto
    ) async throws -> TextToSpeech {
        let inner = try await NobodyWhoGenerated.loadTextToSpeech(
            source: source,
            architecture: architecture?.rawValue,
            voice: voice,
            language: language,
            speed: speed,
            steps: steps,
            silenceDuration: silenceDuration,
            precision: precision,
            temperature: temperature,
            huggingfaceToken: huggingfaceToken,
            device: device.rawValue
        )
        return TextToSpeech(inner: inner)
    }

    /// Synthesize text and return WAV bytes.
    public func synthesize(_ text: String) async throws -> Data {
        try await inner.synthesizeAsync(text: text)
    }

    /// Synthesize text synchronously and return WAV bytes.
    public func synthesizeSync(_ text: String) throws -> Data {
        try inner.synthesize(text: text)
    }
}
